README.md
# color-ral

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar color-ral-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2023 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-07-13
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-07-13
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/color-ral/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/color-ral/compare/0.1.0...0.1.1


test/color_ral/core_test.clj
(ns color-ral.core-test
  (:import (org.openqa.selenium By
                                Capabilities
                                Dimension
                                Keys
                                NoSuchElementException
                                OutputType
                                Point
                                TakesScreenshot
                                WebDriver
                                WebElement
                                NoSuchElementException)
           (org.openqa.selenium.support.ui WebDriverWait
                                           ExpectedConditions)
           org.openqa.selenium.edge.EdgeDriver
           org.openqa.selenium.safari.SafariDriver
           org.openqa.selenium.chrome.ChromeDriver
           org.openqa.selenium.ie.InternetExplorerDriver
           (org.openqa.selenium.firefox FirefoxDriver
                                        FirefoxProfile)
           (org.openqa.selenium.interactions Actions
                                             CompositeAction))
  (:require [clojure.test :refer :all]
            [color-ral.core :refer :all])
  (:gen-class))
(deftest safari-driver
  ;;(org.apache.log4j.BasicConfigurator/configure)
  (with-open [safari (SafariDriver.)]
    (.get safari "http://www.ralcolor.com")
    (testing "detect-patterns"
      (let [ral-1000 "/html/body/div/div/table/tbody/tr[14]/td[2]/span"
            ral-1001 "/html/body/div/div/table/tbody/tr[16]/td[3]/span"
            ral-1002 "/html/body/div/div/table/tbody/tr[18]/td[3]/p/span"
            ral-1003 "/html/body/div/div/table/tbody/tr[20]/td[3]/p/span"
            ral-1004 "/html/body/div/div/table/tbody/tr[22]/td[3]/p/span"
            ral-1005 "/html/body/div/div/table/tbody/tr[24]/td[3]/p/span"
            ral-1018 "/html/body/div/div/table/tbody/tr[44]/td[3]/p/span"
            ral-1037 "/html/body/div/div/table/tbody/tr[72]/td[3]/p/span"
            ral-2000 "/html/body/div/div/table/tbody/tr[74]/td[4]/p/span"
            ral-3000 "/html/body/div/div/table/tbody/tr[100]/td[4]/p/span"
            ral-6007 "/html/body/div/div/table/tbody/tr[238]/td[3]/p/span"
            ral-8028 "/html/body/div/div/table/tbody/tr[408]/td[3]/p/span"
            ral-8029 "/html/body/div/div/table/tbody/tr[410]/td[3]/p/span"
            ral-9001-0 "/html/body/div/div/table/tbody/tr[412]/td[1]"
            ral-9001-1 "/html/body/div/div/table/tbody/tr[412]/td[2]/p/span"
            ral-9001-3 "/html/body/div/div/table/tbody/tr[412]/td[4]/p/span"
            ral-9001-2 "/html/body/div/div/table/tbody/tr[412]/td[3]/p/span"
            ral-9023 "/html/body/div/div/table/tbody/tr[438]/td[1]/p/span"]
        (-> safari
            (WebDriverWait. (java.time.Duration/ofSeconds 5))
            (.until (ExpectedConditions/presenceOfElementLocated (By/xpath ral-9001-0))))
        (println "-------------")
        (println (.getText (.findElement safari (By/xpath ral-9001-0))))
        (println (.getText (.findElement safari (By/xpath ral-9001-1))))
        (println (.getText (.findElement safari (By/xpath ral-9001-2))))
        (println (.getText (.findElement safari (By/xpath ral-9001-3))))
        (println (.getText (.findElement safari (By/xpath ral-1001))))
        (println (.getText (.findElement safari (By/xpath ral-1002))))
        (println "-------------")
        )
      "td[1] needs empty-detection to choose td[3] or td[4], td[2] needs trim")
    (let [prefix "/html/body/div/div/table/tbody/tr["
          mid "]/td["
          suffix "]"
          f #(.getText (.findElement safari (By/xpath %)))
          path #(str prefix %1 mid %2 suffix)
          r #(clojure.string/replace % #"[^\d\w]+|[_]+" "")
          padding-paths (map #(path % 1) (range 14 440 2))
          padding-values (map (comp r f) padding-paths)
          values (map #(if (empty? %) 1 0) padding-values)
          xpaths (map #(path %1 (+ 3 %2)) (range 14 440 2) values)]
      (println (map (comp r f) xpaths)))
    (println (let [path #(str "/html/body/div/div/table/tbody/tr[" %1 "]/td[" %2 "]")
          get-text #(.getText (.findElement safari (By/xpath %)))
          trim #(clojure.string/replace % #"[^\d\w]+|[_]+" "")]
          (for [tab [1 3]]
            (map get-text (map #(path %1 (+ tab %2))
                                  (range 14 440 2)
                                  (map (comp #(if (empty? %) 1 0) trim get-text #(path % 1)) (range 14 440 2)))))))))


src/color_ral/core.clj
(ns color-ral.core
  (:import (org.openqa.selenium By
                                Capabilities
                                Dimension
                                Keys
                                NoSuchElementException
                                OutputType
                                Point
                                TakesScreenshot
                                WebDriver
                                WebElement
                                NoSuchElementException)
           (org.openqa.selenium.support.ui WebDriverWait
                                           ExpectedConditions)
           org.openqa.selenium.edge.EdgeDriver
           org.openqa.selenium.safari.SafariDriver
           org.openqa.selenium.chrome.ChromeDriver
           org.openqa.selenium.ie.InternetExplorerDriver
           (org.openqa.selenium.firefox FirefoxDriver
                                        FirefoxProfile)
           (org.openqa.selenium.interactions Actions
                                             CompositeAction))
  (:gen-class))
(defn- path [index tab] (str "/html/body/div/div/table/tbody/tr[" index "]/td[" tab "]"))
(defn- get-text [^WebDriver driver xpath] (.getText (.findElement driver (By/xpath xpath))))
(defn- trim [str]
  (-> (clojure.string/replace str #"[^\d\w]+|[_]+" " ")
       clojure.string/trim
       (clojure.string/replace #"\s+" "_")))
(defn- get-color [^WebDriver driver xpath]
  (.getCssValue (.findElement driver (By/xpath xpath)) "background-color"))
(defn- serilize [rgb]
  (->> (clojure.string/split rgb #"\D+")
       (filter (comp not empty?))
       (map read-string)
       (map #(Integer/toString % 16))
       (apply str "#FF")
       clojure.string/upper-case))
(defn -main
  [& args]
  (with-open [writer (clojure.java.io/writer "target/colors.xml" :append false)]
    (spit writer "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<resources>\n"))
  (with-open [writer (clojure.java.io/writer "target/colors.xml" :append true)
              safari (SafariDriver.)]
    (.get safari (str "http://www.ralcolor.com"))
    (-> safari
        (WebDriverWait. (java.time.Duration/ofSeconds 5))
        (.until (ExpectedConditions/presenceOfElementLocated (By/xpath (path 412 1)))))
    (doseq [index (range 14 440 2)
            tab [3]]
      (.write writer "    <color name=\"")
      (.write writer (->> (path index 1)
                          (get-text safari)
                          trim
                          empty?
                          (#(if % 1 0))
                          (+ tab)
                          (path index)
                          (get-text safari)
                          trim))
      (.write writer "\">")
      (.write writer (->> (path index tab)
                          (get-color safari)
                          serilize))
      (.write writer "</color>\n"))
    (.write writer "</resources>\n")))
